ls1intum/Artemis
# [Feature] Manual inline feedback for Programming Exercises and fix repo is locked label

### Checklist

### Motivation and Context


At the moment tutors or instructors are not able to give feedback for a specific line of code for programming exercises.

### Description

- Use CodeEditorAceComponent for the inline feedbacks (referenced feedback)
- Created InlineFeedbackComponent which is displayed in the AceEditor
- Allow tutors to add unreferenced and general feedback 
- Add label which indicates how many points the assessment has
- UI fix for 'repository is locked' label in CodeEditorStudentComponent

### Steps for Testing


1. Log in to Artemis
2. Navigate to Course Administration
3. Create a programming exercise and participate with several different users
4. Login as tutor/instructor and assess the programming exercise
5. When hovering over the gutter (line numbers on the left) an plus icon appears and click
6. Add inline feedback
7. Add unreferenced feedback
8. Add general feedback
9. Check if the points label adapts accordingly
10. Submit your assessment and see if everything is updated accordingly and works as expected (check also student view)
11. Check also for complaints and exam

### Test Coverage

* code-editor-tutor-assessment-inline-feedback.component.ts: 95%
* code-editor-tutor-assessment-container.component.ts: 69.49%
* code-editor-ace.component.ts: 66.06%

### Screenshots


As a Tutor:

Hover over code and click on plus button..
![image](https://user-images.githubusercontent.com/41108487/93323075-6b36a880-f814-11ea-87ef-396f60087975.png)

.. inline comment appears and you can enter feedback text and points. Color changes depending on the given points..
![image](https://user-images.githubusercontent.com/41108487/93323161-89040d80-f814-11ea-9359-8b1e0ab54471.png)
..after clicking on save.
![image](https://user-images.githubusercontent.com/41108487/93323190-93260c00-f814-11ea-9540-f51af36de6ad.png)

Unrefenreced and general feedback as known from the other exercises:
![image](https://user-images.githubusercontent.com/41108487/93324073-b1d8d280-f815-11ea-9871-ee09d0a80604.png)



Student view:

<img width="1370" alt="image" src="https://user-images.githubusercontent.com/41108487/93323588-19425280-f815-11ea-850d-cdded18221aa.png">

Feedback info:
<img width="906" alt="image" src="https://user-images.githubusercontent.com/41108487/93323717-4131b600-f815-11ea-9b5e-7728a31b089d.png">
